File: node/packages/botas/src/clients/user-token-client.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { BotHttpClient, type TokenProvider } from './bot-http-client.js'
import type { Activity } from '../schema/activity.js'
import { getLogger } from '../logging/logger.js'

const TOKEN_SERVICE_URL = 'https://token.botframework.com'

/** The OAuth token status for a user on a specific connection. */
export interface TokenStatus {
  channelId: string;
  connectionName: string;
  hasToken: boolean;
  serviceProviderDisplayName?: string;
}

/** An OAuth token for a user. */
export interface UserToken {
  channelId: string;
  connectionName: string;
  token: string;
  expiration?: string;
  properties?: Record<string, unknown>;
}

/** Resource returned by the Bot Framework sign-in endpoint. */
export interface SignInResource {
  signInLink?: string;
  tokenExchangeResource?: TokenExchangeResource;
  tokenPostResource?: TokenPostResource;
}

/** Metadata for an SSO token exchange. */
export interface TokenExchangeResource {
  id: string;
  uri?: string;
  providerId?: string;
}

/** SAS URL for direct token posting. */
export interface TokenPostResource {
  sasUrl?: string;
}

/** Request body for exchanging an SSO token. */
export interface TokenExchangeRequest {
  uri?: string;
  token?: string;
}

/** A set of AAD resource URLs to fetch tokens for. */
export interface AadResourceUrls {
  resourceUrls: string[];
}

/**
 * Client for the Bot Framework token service.
 *
 * Provides methods for OAuth sign-in flows, token retrieval, SSO token
 * exchange, user sign-out, and AAD token acquisition.
 */
export class UserTokenClient {
  private readonly http: BotHttpClient

  constructor (getToken?: TokenProvider) {
    this.http = new BotHttpClient(getToken)
  }

  async getTokenStatusAsync (
    userId: string,
    channelId: string,
    connectionName?: string
  ): Promise<TokenStatus[]> {
    getLogger().info('Calling API endpoint: GetTokenStatus')
    const result = await this.http.get<TokenStatus[]>(
      TOKEN_SERVICE_URL,
      '/api/usertoken/GetTokenStatus',
      { userId, channelId, include: connectionName },
      { operationDescription: 'get token status' }
    )
    return result ?? []
  }

  async getTokenAsync (
    userId: string,
    channelId: string,
    connectionName: string,
    code?: string
  ): Promise<UserToken | undefined> {
    getLogger().info('Calling API endpoint: GetToken')
    return this.http.get<UserToken>(
      TOKEN_SERVICE_URL,
      '/api/usertoken/GetToken',
      { userId, channelId, connectionName, code },
      { operationDescription: 'get token', returnNullOnNotFound: true }
    )
  }

  async getSignInResourceAsync (
    connectionName: string,
    activity: Activity,
    finalRedirect?: string
  ): Promise<SignInResource | undefined> {
    getLogger().info('Calling API endpoint: GetSignInResource')
    return this.http.get<SignInResource>(
      TOKEN_SERVICE_URL,
      '/api/botsignin/GetSignInResource',
      { state: buildStateParam(connectionName, activity), finalRedirect },
      { operationDescription: 'get sign-in resource' }
    )
  }

  async exchangeTokenAsync (
    userId: string,
    channelId: string,
    connectionName: string,
    request: TokenExchangeRequest
  ): Promise<UserToken | undefined> {
    getLogger().info('Calling API endpoint: ExchangeToken')
    const params = { userId, channelId, connectionName }
    const query = Object.entries(params)
      .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
      .join('&')
    const url = `${TOKEN_SERVICE_URL}/api/usertoken/ExchangeToken?${query}`
    return this.http.send<UserToken>('POST', url, request, {
      operationDescription: 'exchange token',
    })
  }

  async signOutUserAsync (
    userId: string,
    channelId: string,
    connectionName?: string
  ): Promise<void> {
    getLogger().info('Calling API endpoint: SignOut')
    await this.http.delete(
      TOKEN_SERVICE_URL,
      '/api/usertoken/SignOut',
      { userId, channelId, connectionName },
      { operationDescription: 'sign out user' }
    )
  }

  async getAadTokensAsync (
    userId: string,
    channelId: string,
    connectionName: string,
    resourceUrls: AadResourceUrls
  ): Promise<Record<string, UserToken>> {
    getLogger().info('Calling API endpoint: GetAadTokens')
    const params = { userId, channelId, connectionName }
    const query = Object.entries(params)
      .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
      .join('&')
    const url = `${TOKEN_SERVICE_URL}/api/usertoken/GetAadTokens?${query}`
    const result = await this.http.send<Record<string, UserToken>>(
      'POST',
      url,
      resourceUrls,
      { operationDescription: 'get AAD tokens' }
    )
    return result ?? {}
  }
}

function buildStateParam (connectionName: string, activity: Activity): string {
  const state = {
    ConnectionName: connectionName,
    Conversation: {
      ActivityId: activity.id,
      Bot: activity.recipient,
      ChannelId: activity.channelId,
      Conversation: activity.conversation,
      ServiceUrl: activity.serviceUrl,
    },
    RelatesTo: null,
    MSAppId: activity.recipient?.id,
  }
  return Buffer.from(JSON.stringify(state)).toString('base64')
}

File: python/packages/botas/src/botas/clients/user_token_client.py
from __future__ import annotations

import base64
import json
from typing import Any

from botas.clients.bot_http_client import BotHttpClient, BotRequestOptions, TokenProvider
from botas.schema.core_activity import CoreActivity, _CamelModel

_TOKEN_SERVICE_URL = "https://token.botframework.com"


class TokenStatus(_CamelModel):
    channel_id: str
    connection_name: str
    has_token: bool = False
    service_provider_display_name: str | None = None


class UserToken(_CamelModel):
    connection_name: str
    token: str
    expiration: str | None = None


class SignInResource(_CamelModel):
    sign_in_link: str
    token_exchange_resource: dict[str, Any] | None = None


class TokenExchangeRequest(_CamelModel):
    uri: str | None = None
    token: str | None = None


class UserTokenClient:
    def __init__(self, get_token: TokenProvider | None = None) -> None:
        self._http = BotHttpClient(get_token)

    async def get_token_status_async(
        self, user_id: str, channel_id: str, connection_name: str | None = None
    ) -> list[TokenStatus]:
        params: dict[str, str | None] = {
            "userId": user_id,
            "channelId": channel_id,
            "include": connection_name,
        }
        data = await self._http.get(
            _TOKEN_SERVICE_URL, "/api/usertoken/GetTokenStatus", params=params,
            options=BotRequestOptions(operation_description="get token status"),
        )
        return [TokenStatus.model_validate(t) for t in (data or [])]

    async def get_token_async(
        self,
        user_id: str,
        channel_id: str,
        connection_name: str,
        code: str | None = None,
    ) -> UserToken | None:
        params: dict[str, str | None] = {
            "userId": user_id,
            "channelId": channel_id,
            "connectionName": connection_name,
            "code": code,
        }
        data = await self._http.get(
            _TOKEN_SERVICE_URL, "/api/usertoken/GetToken", params=params,
            options=BotRequestOptions(operation_description="get token", return_none_on_not_found=True),
        )
        return UserToken.model_validate(data) if data else None

    async def get_sign_in_resource_async(
        self,
        connection_name: str,
        activity: CoreActivity,
        final_redirect: str | None = None,
    ) -> SignInResource | None:
        state = self._build_state(connection_name, activity)
        params: dict[str, str | None] = {
            "state": state,
            "finalRedirect": final_redirect,
        }
        data = await self._http.get(
            _TOKEN_SERVICE_URL, "/api/botsignin/GetSignInResource", params=params,
            options=BotRequestOptions(operation_description="get sign-in resource"),
        )
        return SignInResource.model_validate(data) if data else None

    async def exchange_token_async(
        self,
        user_id: str,
        channel_id: str,
        connection_name: str,
        request: TokenExchangeRequest,
    ) -> UserToken | None:
        params: dict[str, str | None] = {
            "userId": user_id,
            "channelId": channel_id,
            "connectionName": connection_name,
        }
        data = await self._http.post(
            _TOKEN_SERVICE_URL,
            "/api/usertoken/exchange?" + "&".join(f"{k}={v}" for k, v in params.items() if v),
            request.model_dump(by_alias=True, exclude_none=True),
            BotRequestOptions(operation_description="exchange token"),
        )
        return UserToken.model_validate(data) if data else None

    async def sign_out_user_async(
        self, user_id: str, channel_id: str, connection_name: str | None = None
    ) -> None:
        params: dict[str, str | None] = {
            "userId": user_id,
            "channelId": channel_id,
            "connectionName": connection_name,
        }
        endpoint = "/api/usertoken/SignOut"
        query = "&".join(f"{k}={v}" for k, v in params.items() if v is not None)
        if query:
            endpoint = f"{endpoint}?{query}"
        await self._http.delete(
            _TOKEN_SERVICE_URL, endpoint,
            BotRequestOptions(operation_description="sign out user"),
        )

    def _build_state(self, connection_name: str, activity: CoreActivity) -> str:
        ext = activity.model_extra or {}
        state = {
            "connectionName": connection_name,
            "conversation": {
                "id": activity.conversation.id if activity.conversation else "",
                "channelId": ext.get("channelId"),
                "serviceUrl": activity.service_url,
            },
            "relatesTo": ext.get("id"),
        }
        return base64.b64encode(json.dumps(state).encode()).decode()
